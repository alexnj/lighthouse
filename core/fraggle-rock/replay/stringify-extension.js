/**
 * @license Copyright 2022 The Lighthouse Authors. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */
'use strict';

import * as PuppeteerReplay from '@puppeteer/replay';

/**
 * @param {PuppeteerReplay.Schema.Step} step
 * @return {boolean}
 */
function isNavigationStep(step) {
  return Boolean(
    step.type === 'navigate' ||
    step.assertedEvents?.some(event => event.type === 'navigation')
  );
}

class LighthouseStringifyExtension extends PuppeteerReplay.PuppeteerStringifyExtension {
  #isProcessingTimespan = false;

  /**
   * @override
   * @param {PuppeteerReplay.LineWriter} out
   * @param {PuppeteerReplay.Schema.UserFlow} flow
   */
  async beforeAllSteps(out, flow) {
    out.appendLine(`const fs = require('fs');`);

    let isMobile = true;
    for (const step of flow.steps) {
      if (step.type !== 'setViewport') continue;
      isMobile = step.isMobile;
    }

    await super.beforeAllSteps(out, flow);

    const flags = {
      screenEmulation: {
        disabled: true,
      },
    };
    out.appendLine(`const flags = ${JSON.stringify(flags)}`);
    if (isMobile) {
      out.appendLine(`const config = undefined;`);
    } else {
      // eslint-disable-next-line max-len
      out.appendLine(`const config = (await import('lighthouse/core/config/desktop-config.js')).default;`);
    }

    out.appendLine(`const lhApi = await import('lighthouse/core/fraggle-rock/api.js');`);
    // eslint-disable-next-line max-len
    out.appendLine(`const lhFlow = await lhApi.startFlow(page, {name: ${JSON.stringify(flow.title)}, config, flags});`);
  }

  /**
   * @override
   * @param {PuppeteerReplay.LineWriter} out
   * @param {PuppeteerReplay.Schema.Step} step
   * @param {PuppeteerReplay.Schema.UserFlow} flow
   */
  async stringifyStep(out, step, flow) {
    if (step.type === 'setViewport') {
      await super.stringifyStep(out, step, flow);
      return;
    }

    const isNavigation = isNavigationStep(step);

    if (isNavigation) {
      if (this.#isProcessingTimespan) {
        out.appendLine(`await lhFlow.endTimespan();`);
        this.#isProcessingTimespan = false;
      }
      out.appendLine(`await lhFlow.startNavigation();`);
    } else if (!this.#isProcessingTimespan) {
      out.appendLine(`await lhFlow.startTimespan();`);
      this.#isProcessingTimespan = true;
    }

    await super.stringifyStep(out, step, flow);

    if (isNavigation) {
      out.appendLine(`await lhFlow.endNavigation();`);
    }
  }

  /**
   * @override
   * @param {PuppeteerReplay.LineWriter} out
   * @param {PuppeteerReplay.Schema.UserFlow} flow
   */
  async afterAllSteps(out, flow) {
    if (this.#isProcessingTimespan) {
      out.appendLine(`await lhFlow.endTimespan();`);
    }
    out.appendLine(`const lhFlowReport = await lhFlow.generateReport();`);
    out.appendLine(`fs.writeFileSync(__dirname + '/flow.report.html', lhFlowReport)`);
    await super.afterAllSteps(out, flow);
  }
}

export default LighthouseStringifyExtension;
