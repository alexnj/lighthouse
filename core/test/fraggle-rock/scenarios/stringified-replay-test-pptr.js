/**
 * @license Copyright 2022 The Lighthouse Authors. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */

import fs from 'fs';
import {promisify} from 'util';
import {execFile} from 'child_process';

import {stringify} from '@puppeteer/replay';

import {LH_ROOT} from '../../../../root.js';
import LighthouseStringifyExtension from '../../../fraggle-rock/replay/stringify-extension.js';
import {getAuditsBreakdown, createTestState} from './pptr-test-utils.js';
import {readJson} from '../../test-utils.js';

const execFileAsync = promisify(execFile);
const desktopReplayJson = readJson('core/test/fixtures/fraggle-rock/replay/desktop-test-flow.json');
const mobileReplayJson = readJson('core/test/fixtures/fraggle-rock/replay/mobile-test-flow.json');
const FLOW_JSON_REGEX = /window\.__LIGHTHOUSE_FLOW_JSON__ = (.*);<\/script>/;

describe('Running the stringified output script', function() {
  // eslint-disable-next-line no-invalid-this
  this.timeout(60_000);

  // Flow JSON specifies port 10200 so we have to use that for the server.
  const state = createTestState();
  state.installServerHooks(10200);

  const tmpDir = `${LH_ROOT}/.tmp/replay`;
  let testTmpDir = '';
  let scriptPath = '';

  before(() => {
    fs.mkdirSync(tmpDir, {recursive: true});
  });

  beforeEach(() => {
    testTmpDir = fs.mkdtempSync(`${tmpDir}/replay-`);
    scriptPath = `${testTmpDir}/stringified.cjs`;
  });

  after(() => {
    fs.rmSync(tmpDir, {recursive: true, force: true});
  });

  it('generates a valid desktop flow report', async () => {
    const scriptContents = await stringify(desktopReplayJson, {
      extension: new LighthouseStringifyExtension(),
    });

    expect(scriptContents).toMatchSnapshot();
    fs.writeFileSync(scriptPath, scriptContents);

    const {stdout, stderr} = await execFileAsync('node', [scriptPath], {timeout: 50_000});

    // Ensure script didn't quietly report an issue.
    expect(stdout).toEqual('');
    expect(stderr).toEqual('');

    const reportHtml = fs.readFileSync(`${testTmpDir}/flow.report.html`, 'utf-8');
    const flowResultJson = FLOW_JSON_REGEX.exec(reportHtml)?.[1];
    if (!flowResultJson) throw new Error('Could not find flow json');

    /** @type {LH.FlowResult} */
    const flowResult = JSON.parse(flowResultJson);
    expect(flowResult.name).toEqual(desktopReplayJson.title);
    expect(flowResult.steps.map(step => step.lhr.gatherMode)).toEqual([
      'navigation',
      'timespan',
      'navigation',
      'timespan',
    ]);

    for (const {lhr} of flowResult.steps) {
      expect(lhr.configSettings.formFactor).toEqual('desktop');
      expect(lhr.configSettings.screenEmulation.disabled).toBeTruthy();

      const {auditResults, erroredAudits} = getAuditsBreakdown(lhr);
      expect(auditResults.length).toBeGreaterThanOrEqual(10);
      expect(erroredAudits.length).toStrictEqual(0);
    }
  });

  it('generates a valid mobile flow report', async () => {
    const scriptContents = await stringify(mobileReplayJson, {
      extension: new LighthouseStringifyExtension(),
    });

    expect(scriptContents).toMatchSnapshot();
    fs.writeFileSync(scriptPath, scriptContents);

    const {stdout, stderr} = await execFileAsync('node', [scriptPath], {timeout: 50_000});

    // Ensure script didn't quietly report an issue.
    expect(stdout).toEqual('');
    expect(stderr).toEqual('');

    const reportHtml = fs.readFileSync(`${testTmpDir}/flow.report.html`, 'utf-8');
    const flowResultJson = FLOW_JSON_REGEX.exec(reportHtml)?.[1];
    if (!flowResultJson) throw new Error('Could not find flow json');

    /** @type {LH.FlowResult} */
    const flowResult = JSON.parse(flowResultJson);
    expect(flowResult.name).toEqual(mobileReplayJson.title);
    expect(flowResult.steps.map(step => step.lhr.gatherMode)).toEqual([
      'navigation',
      'timespan',
      'navigation',
      'timespan',
    ]);

    for (const {lhr} of flowResult.steps) {
      expect(lhr.configSettings.formFactor).toEqual('mobile');
      expect(lhr.configSettings.screenEmulation.disabled).toBeTruthy();

      const {auditResults, erroredAudits} = getAuditsBreakdown(lhr);
      expect(auditResults.length).toBeGreaterThanOrEqual(10);
      expect(erroredAudits.length).toStrictEqual(0);
    }
  });
});
