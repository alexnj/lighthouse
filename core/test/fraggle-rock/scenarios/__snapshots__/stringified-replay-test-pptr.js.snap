// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Running the stringified output script generates a valid desktop flow report 1`] = `
"const fs = require('fs');
const puppeteer = require('puppeteer'); // v13.0.0 or later

(async () => {
  const browser = await puppeteer.launch();
  const page = await browser.newPage();
  const timeout = 5000;
  page.setDefaultTimeout(timeout);

  const flags = {\\"screenEmulation\\":{\\"disabled\\":true}}
  const config = (await import('lighthouse/core/config/desktop-config.js')).default;
  const lhApi = await import('lighthouse/core/fraggle-rock/api.js');
  const lhFlow = await lhApi.startFlow(page, {name: \\"Test Flow on Desktop\\", config, flags});
  {
    const targetPage = page;
    await targetPage.setViewport({\\"width\\":757,\\"height\\":988})
  }
  await lhFlow.startNavigation();
  {
    const targetPage = page;
    const promises = [];
    promises.push(targetPage.waitForNavigation());
    await targetPage.goto(\\"http://localhost:10200/flow/index.html\\");
    await Promise.all(promises);
  }
  await lhFlow.endNavigation();
  await lhFlow.startTimespan();
  {
    const targetPage = page;
    const element = await waitForSelectors([[\\"aria/Add item\\"],[\\"#add-item\\"]], targetPage, { timeout, visible: true });
    await scrollIntoViewIfNeeded(element, timeout);
    await element.click({
      offset: {
        x: 61,
        y: 13.5625,
      },
    });
  }
  {
    const targetPage = page;
    const element = await waitForSelectors([[\\"aria/Add item\\"],[\\"#add-item\\"]], targetPage, { timeout, visible: true });
    await scrollIntoViewIfNeeded(element, timeout);
    await element.click({
      offset: {
        x: 42,
        y: 4.125,
      },
    });
  }
  {
    const targetPage = page;
    const element = await waitForSelectors([[\\"aria/Add item\\"],[\\"#add-item\\"]], targetPage, { timeout, visible: true });
    await scrollIntoViewIfNeeded(element, timeout);
    await element.click({
      offset: {
        x: 50,
        y: 8.40625,
      },
    });
  }
  await lhFlow.endTimespan();
  await lhFlow.startNavigation();
  {
    const targetPage = page;
    const promises = [];
    promises.push(targetPage.waitForNavigation());
    const element = await waitForSelectors([[\\"aria/Link to new page!\\"],[\\"#link\\"]], targetPage, { timeout, visible: true });
    await scrollIntoViewIfNeeded(element, timeout);
    await element.click({
      offset: {
        x: 53.390625,
        y: 3.6875,
      },
    });
    await Promise.all(promises);
  }
  await lhFlow.endNavigation();
  await lhFlow.startTimespan();
  {
    const targetPage = page;
    const element = await waitForSelectors([[\\"aria/Hidden content here:\\"],[\\"#content > summary\\"]], targetPage, { timeout, visible: true });
    await scrollIntoViewIfNeeded(element, timeout);
    await element.click({
      offset: {
        x: 6,
        y: 10.5625,
      },
    });
  }
  await lhFlow.endTimespan();
  const lhFlowReport = await lhFlow.generateReport();
  fs.writeFileSync(__dirname + '/flow.report.html', lhFlowReport)

  await browser.close();

  async function waitForSelectors(selectors, frame, options) {
    for (const selector of selectors) {
      try {
        return await waitForSelector(selector, frame, options);
      } catch (err) {
        console.error(err);
      }
    }
    throw new Error('Could not find element for selectors: ' + JSON.stringify(selectors));
  }

  async function scrollIntoViewIfNeeded(element, timeout) {
    await waitForConnected(element, timeout);
    const isInViewport = await element.isIntersectingViewport({threshold: 0});
    if (isInViewport) {
      return;
    }
    await element.evaluate(element => {
      element.scrollIntoView({
        block: 'center',
        inline: 'center',
        behavior: 'auto',
      });
    });
    await waitForInViewport(element, timeout);
  }

  async function waitForConnected(element, timeout) {
    await waitForFunction(async () => {
      return await element.getProperty('isConnected');
    }, timeout);
  }

  async function waitForInViewport(element, timeout) {
    await waitForFunction(async () => {
      return await element.isIntersectingViewport({threshold: 0});
    }, timeout);
  }

  async function waitForSelector(selector, frame, options) {
    if (!Array.isArray(selector)) {
      selector = [selector];
    }
    if (!selector.length) {
      throw new Error('Empty selector provided to waitForSelector');
    }
    let element = null;
    for (let i = 0; i < selector.length; i++) {
      const part = selector[i];
      if (element) {
        element = await element.waitForSelector(part, options);
      } else {
        element = await frame.waitForSelector(part, options);
      }
      if (!element) {
        throw new Error('Could not find element: ' + selector.join('>>'));
      }
      if (i < selector.length - 1) {
        element = (await element.evaluateHandle(el => el.shadowRoot ? el.shadowRoot : el)).asElement();
      }
    }
    if (!element) {
      throw new Error('Could not find element: ' + selector.join('|'));
    }
    return element;
  }

  async function waitForElement(step, frame, timeout) {
    const count = step.count || 1;
    const operator = step.operator || '>=';
    const comp = {
      '==': (a, b) => a === b,
      '>=': (a, b) => a >= b,
      '<=': (a, b) => a <= b,
    };
    const compFn = comp[operator];
    await waitForFunction(async () => {
      const elements = await querySelectorsAll(step.selectors, frame);
      return compFn(elements.length, count);
    }, timeout);
  }

  async function querySelectorsAll(selectors, frame) {
    for (const selector of selectors) {
      const result = await querySelectorAll(selector, frame);
      if (result.length) {
        return result;
      }
    }
    return [];
  }

  async function querySelectorAll(selector, frame) {
    if (!Array.isArray(selector)) {
      selector = [selector];
    }
    if (!selector.length) {
      throw new Error('Empty selector provided to querySelectorAll');
    }
    let elements = [];
    for (let i = 0; i < selector.length; i++) {
      const part = selector[i];
      if (i === 0) {
        elements = await frame.$$(part);
      } else {
        const tmpElements = elements;
        elements = [];
        for (const el of tmpElements) {
          elements.push(...(await el.$$(part)));
        }
      }
      if (elements.length === 0) {
        return [];
      }
      if (i < selector.length - 1) {
        const tmpElements = [];
        for (const el of elements) {
          const newEl = (await el.evaluateHandle(el => el.shadowRoot ? el.shadowRoot : el)).asElement();
          if (newEl) {
            tmpElements.push(newEl);
          }
        }
        elements = tmpElements;
      }
    }
    return elements;
  }

  async function waitForFunction(fn, timeout) {
    let isActive = true;
    setTimeout(() => {
      isActive = false;
    }, timeout);
    while (isActive) {
      const result = await fn();
      if (result) {
        return;
      }
      await new Promise(resolve => setTimeout(resolve, 100));
    }
    throw new Error('Timed out');
  }
})();
"
`;

exports[`Running the stringified output script generates a valid mobile flow report 1`] = `
"const fs = require('fs');
const puppeteer = require('puppeteer'); // v13.0.0 or later

(async () => {
  const browser = await puppeteer.launch();
  const page = await browser.newPage();
  const timeout = 5000;
  page.setDefaultTimeout(timeout);

  const flags = {\\"screenEmulation\\":{\\"disabled\\":true}}
  const config = undefined;
  const lhApi = await import('lighthouse/core/fraggle-rock/api.js');
  const lhFlow = await lhApi.startFlow(page, {name: \\"Test Flow on Mobile\\", config, flags});
  {
    const targetPage = page;
    await targetPage.setViewport({\\"width\\":300,\\"height\\":600})
  }
  await lhFlow.startNavigation();
  {
    const targetPage = page;
    const promises = [];
    promises.push(targetPage.waitForNavigation());
    await targetPage.goto(\\"http://localhost:10200/flow/index.html\\");
    await Promise.all(promises);
  }
  await lhFlow.endNavigation();
  await lhFlow.startTimespan();
  {
    const targetPage = page;
    const element = await waitForSelectors([[\\"aria/Add item\\"],[\\"#add-item\\"]], targetPage, { timeout, visible: true });
    await scrollIntoViewIfNeeded(element, timeout);
    await element.click({
      offset: {
        x: 61,
        y: 13.5625,
      },
    });
  }
  {
    const targetPage = page;
    const element = await waitForSelectors([[\\"aria/Add item\\"],[\\"#add-item\\"]], targetPage, { timeout, visible: true });
    await scrollIntoViewIfNeeded(element, timeout);
    await element.click({
      offset: {
        x: 42,
        y: 4.125,
      },
    });
  }
  {
    const targetPage = page;
    const element = await waitForSelectors([[\\"aria/Add item\\"],[\\"#add-item\\"]], targetPage, { timeout, visible: true });
    await scrollIntoViewIfNeeded(element, timeout);
    await element.click({
      offset: {
        x: 50,
        y: 8.40625,
      },
    });
  }
  await lhFlow.endTimespan();
  await lhFlow.startNavigation();
  {
    const targetPage = page;
    const promises = [];
    promises.push(targetPage.waitForNavigation());
    const element = await waitForSelectors([[\\"aria/Link to new page!\\"],[\\"#link\\"]], targetPage, { timeout, visible: true });
    await scrollIntoViewIfNeeded(element, timeout);
    await element.click({
      offset: {
        x: 53.390625,
        y: 3.6875,
      },
    });
    await Promise.all(promises);
  }
  await lhFlow.endNavigation();
  await lhFlow.startTimespan();
  {
    const targetPage = page;
    const element = await waitForSelectors([[\\"aria/Hidden content here:\\"],[\\"#content > summary\\"]], targetPage, { timeout, visible: true });
    await scrollIntoViewIfNeeded(element, timeout);
    await element.click({
      offset: {
        x: 6,
        y: 10.5625,
      },
    });
  }
  await lhFlow.endTimespan();
  const lhFlowReport = await lhFlow.generateReport();
  fs.writeFileSync(__dirname + '/flow.report.html', lhFlowReport)

  await browser.close();

  async function waitForSelectors(selectors, frame, options) {
    for (const selector of selectors) {
      try {
        return await waitForSelector(selector, frame, options);
      } catch (err) {
        console.error(err);
      }
    }
    throw new Error('Could not find element for selectors: ' + JSON.stringify(selectors));
  }

  async function scrollIntoViewIfNeeded(element, timeout) {
    await waitForConnected(element, timeout);
    const isInViewport = await element.isIntersectingViewport({threshold: 0});
    if (isInViewport) {
      return;
    }
    await element.evaluate(element => {
      element.scrollIntoView({
        block: 'center',
        inline: 'center',
        behavior: 'auto',
      });
    });
    await waitForInViewport(element, timeout);
  }

  async function waitForConnected(element, timeout) {
    await waitForFunction(async () => {
      return await element.getProperty('isConnected');
    }, timeout);
  }

  async function waitForInViewport(element, timeout) {
    await waitForFunction(async () => {
      return await element.isIntersectingViewport({threshold: 0});
    }, timeout);
  }

  async function waitForSelector(selector, frame, options) {
    if (!Array.isArray(selector)) {
      selector = [selector];
    }
    if (!selector.length) {
      throw new Error('Empty selector provided to waitForSelector');
    }
    let element = null;
    for (let i = 0; i < selector.length; i++) {
      const part = selector[i];
      if (element) {
        element = await element.waitForSelector(part, options);
      } else {
        element = await frame.waitForSelector(part, options);
      }
      if (!element) {
        throw new Error('Could not find element: ' + selector.join('>>'));
      }
      if (i < selector.length - 1) {
        element = (await element.evaluateHandle(el => el.shadowRoot ? el.shadowRoot : el)).asElement();
      }
    }
    if (!element) {
      throw new Error('Could not find element: ' + selector.join('|'));
    }
    return element;
  }

  async function waitForElement(step, frame, timeout) {
    const count = step.count || 1;
    const operator = step.operator || '>=';
    const comp = {
      '==': (a, b) => a === b,
      '>=': (a, b) => a >= b,
      '<=': (a, b) => a <= b,
    };
    const compFn = comp[operator];
    await waitForFunction(async () => {
      const elements = await querySelectorsAll(step.selectors, frame);
      return compFn(elements.length, count);
    }, timeout);
  }

  async function querySelectorsAll(selectors, frame) {
    for (const selector of selectors) {
      const result = await querySelectorAll(selector, frame);
      if (result.length) {
        return result;
      }
    }
    return [];
  }

  async function querySelectorAll(selector, frame) {
    if (!Array.isArray(selector)) {
      selector = [selector];
    }
    if (!selector.length) {
      throw new Error('Empty selector provided to querySelectorAll');
    }
    let elements = [];
    for (let i = 0; i < selector.length; i++) {
      const part = selector[i];
      if (i === 0) {
        elements = await frame.$$(part);
      } else {
        const tmpElements = elements;
        elements = [];
        for (const el of tmpElements) {
          elements.push(...(await el.$$(part)));
        }
      }
      if (elements.length === 0) {
        return [];
      }
      if (i < selector.length - 1) {
        const tmpElements = [];
        for (const el of elements) {
          const newEl = (await el.evaluateHandle(el => el.shadowRoot ? el.shadowRoot : el)).asElement();
          if (newEl) {
            tmpElements.push(newEl);
          }
        }
        elements = tmpElements;
      }
    }
    return elements;
  }

  async function waitForFunction(fn, timeout) {
    let isActive = true;
    setTimeout(() => {
      isActive = false;
    }, timeout);
    while (isActive) {
      const result = await fn();
      if (result) {
        return;
      }
      await new Promise(resolve => setTimeout(resolve, 100));
    }
    throw new Error('Timed out');
  }
})();
"
`;
